test/arithmetic/adder_test.dart

```diff
@@ -0,0 +1,339 @@
+}
+
+void testExhaustive(int n, Adder Function(Logic a, Logic b) fn) {
+  final a = Logic(name: 'a', width: n);
+  final b = Logic(name: 'b', width: n);
+
+  final mod = fn(a, b);
+  test(
+      'exhaustive: ${mod.name}_W$n'
+      '_G${fn.call(a, b).name}', () async {
+    await mod.build();
+
+    for (var aa = 0; aa < (1 << n); ++aa) {
+      for (var bb = 0; bb < (1 << n); ++bb) {
+        final av = LogicValue.of(BigInt.from(aa), width: n);
+        final bv = LogicValue.of(BigInt.from(bb), width: n);
+        checkAdder(mod, av, bv);
+      }
+    }
+  });
+}
+
+void testAdderRandom(int n, int nSamples, Adder Function(Logic a, Logic b) fn) {
+  final a = Logic(name: 'a', width: n);
+  final b = Logic(name: 'b', width: n);
+
+  final adder = fn(a, b);
+  test('random: ${adder.name}_W${a.width}_I$nSamples', () async {
+    await adder.build();
+
+    for (var i = 0; i < nSamples; ++i) {
+      final aa = Random().nextLogicValue(width: n);
+      final bb = Random().nextLogicValue(width: n);
+      checkAdder(adder, aa, bb);
+    }
+  });
+}
+
+void checkSignMagnitudeAdder(SignMagnitudeAdder adder, LogicValue aSign,
+    LogicValue aMagnitude, LogicValue bSign, LogicValue bMagnitude) {
+  // ignore: invalid_use_of_protected_member
+  adder.aSign.put(aSign);
+  // ignore: invalid_use_of_protected_member
+  adder.bSign.put(bSign);
+
+  // ignore: invalid_use_of_protected_member
+  adder.a.put(aMagnitude);
+  // ignore: invalid_use_of_protected_member
+  adder.b.put(bMagnitude);
+
+  final computedVal = (adder.sign.value == LogicValue.one)
+      ? -adder.sum.value.toBigInt()
+      : adder.sum.value.toBigInt();
+
+  final aValue = (aSign == LogicValue.one)
+      ? -aMagnitude.toBigInt()
+      : aMagnitude.toBigInt();
+  final bValue = (bSign == LogicValue.one)
+      ? -bMagnitude.toBigInt()
+      : bMagnitude.toBigInt();
+  final expectSign = (aValue + bValue).sign;
+  final expectedMag = (aValue + bValue).abs();
+
+  final expectVal = expectSign == -1 ? -expectedMag : expectedMag;
+  expect(computedVal, equals(expectVal));
+}
+
+void testExhaustiveSignMagnitude(int n, Adder Function(Logic a, Logic b) fn,
+    {bool sortOperands = true}) {
+  final aSign = Logic(name: 'aSign');
+  final a = Logic(name: 'a', width: n);
+  final bSign = Logic(name: 'bSign');
+  final b = Logic(name: 'b', width: n);
+
+  final adder = SignMagnitudeAdder(aSign, a, bSign, b, fn,
+      largestMagnitudeFirst: sortOperands);
+  test('exhaustive Sign Magnitude: ${adder.name}_W${a.width}_N$sortOperands',
+      () {
+    for (var i = 0; i < pow(2, n); i += 1) {
+      for (var j = 0; j < pow(2, n); j += 1) {
+        final bI = BigInt.from(i).toSigned(n);
+        final bJ = BigInt.from(j).toSigned(n);
+
+        final bigger = bI;
+        final smaller = bJ;
+        // When equal, we want the negative one first to produce 1 1...1
+        if (sortOperands &
+            ((bI.abs() < bJ.abs()) || (bI.abs() == bJ.abs() && (bI > bJ)))) {
+          continue;
```

Comment:
Is this not filtering it such that we don't test the ability for the hardware to handle sorting the operands itself?

Reply:
If sortOperands is false, then the hardware is doing the sorting.  If sortOperands is true, we need to continue past this case as the operands are in bad order.

Reply:
I renamed to operandsArePreSorted

Reply:
ahh i see, thanks!